File: leetcode-js/329-longest-increasing-path-in-a-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number}
 */
var longestIncreasingPath = function(matrix) {
    if (!matrix.length) {
        return 0;
    }
    const cache = Array(matrix.length).fill().map(() => Array(matrix[0].length).fill(0));
    let max = 1;
    for (let i = 0; i < matrix.length; i += 1) {
        for (let j = 0; j < matrix[0].length; j += 1) {
            const len = dfs(matrix, i, j, matrix.length, matrix[0].length, cache);
            max = Math.max(max, len);
        }
    }
    return max;
};

const DIRS = [
    [0, 1],
    [1, 0],
    [0, -1],
    [-1, 0]
];

function dfs(matrix, i, j, m, n, cache) {
    if (cache[i][j] !== 0) {
        return cache[i][j];
    }
    let max = 1;
    for (const dir of DIRS) {
        const x = i + dir[0];
        const y = j + dir[1];
        if (x < 0 || x >= m || y < 0 || y >= n || matrix[x][y] <= matrix[i][j]) {
            continue;
        }
        const len = 1 + dfs(matrix, x, y, m, n, cache);
        max = Math.max(max, len);
    }
    cache[i][j] = max;
    return max;
}

console.log(longestIncreasingPath([[9,9,4],[6,6,8],[2,1,1]]));

File: leetcode-java/329-longest-increasing-path-in-a-matrix.java
public class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        if (matrix.length == 0) {
            return 0;
        }

        int m = matrix.length;
        int n = matrix[0].length;
        int[][] distances = new int[m][n];
        int max = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                max = Math.max(max, dfs(matrix, m, n, distances, i, j, Integer.MIN_VALUE));
            }
        }

        return max;
    }

    int dfs(int[][] matrix, int m, int n, int[][] distances, int i, int j, int curr) {
        if (i < 0 || i >= m || j < 0 || j >= n || curr >= matrix[i][j]) {
            return 0;
        }

        if (distances[i][j] != 0) {
            return distances[i][j];
        }

        int max = 0;

        max = Math.max(max, dfs(matrix, m, n, distances, i - 1, j, matrix[i][j]));
        max = Math.max(max, dfs(matrix, m, n, distances, i + 1, j, matrix[i][j]));
        max = Math.max(max, dfs(matrix, m, n, distances, i, j - 1, matrix[i][j]));
        max = Math.max(max, dfs(matrix, m, n, distances, i, j + 1, matrix[i][j]));

        distances[i][j] = max + 1;

        return max + 1;
    }
}
